app/controllers/concerns/course_basic_detail_concern.rb

```diff
@@ -89,18 +104,63 @@ def build_previous_course_creation_params
   end
 
   def next_step(current_step:)
-    if current_step == :outcome
-      new_provider_recruitment_cycle_courses_entry_requirements_path(
+    next_step = NextCourseCreationStepService.new.execute(current_step: current_step)
+
+    if course_creation_paths.key?(next_step)
+      course_creation_paths[next_step]
+    else
+      raise "No path defined for next step: #{next_step}"
+    end
+  end
+
+  def course_creation_paths
+    {
+      apprenticeship:  new_provider_recruitment_cycle_courses_apprenticeship_path(
         params[:provider_code],
         params[:recruitment_cycle_year],
         course: course_params,
-      )
-    elsif current_step == :entry_requirements
-      new_provider_recruitment_cycle_courses_outcome_path(
+      ),
+      # Currently the page isnt built - so skip
+      location: new_provider_recruitment_cycle_courses_entry_requirements_path(
         params[:provider_code],
         params[:recruitment_cycle_year],
         course: course_params,
-      )
-    end
+      ),
+      entry_requirements: new_provider_recruitment_cycle_courses_entry_requirements_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+      outcome: new_provider_recruitment_cycle_courses_outcome_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+      full_or_part_time: new_provider_recruitment_cycle_courses_study_mode_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+      applications_open: new_provider_recruitment_cycle_courses_applications_open_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+      start_date: new_provider_recruitment_cycle_courses_start_date_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+      age_range: new_provider_recruitment_cycle_courses_age_range_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+      confirmation: confirmation_provider_recruitment_cycle_courses_path(
+        params[:provider_code],
+        params[:recruitment_cycle_year],
+        course: course_params,
+      ),
+    }
```

Comment:
I have this niggling feeling about this. Basically we're generating N number of paths for this method when we call it. I have a feeling that generating paths is slow, and generating a bunch of these when we don't need them is maybe worth avoiding. Looking at benchmarks, it's not the worst thing in the world, but cumulatively this could add up to a slow service. Running this 100 times takes 0.33s, vs 0.0002s for NextCourseCreationStepService.new.execute.

There's two ways we could improve this:
- create a method per path, but then we'd have to do `__send__("path_for_#{next_step}")`
- use a case when in here, which leads to a slightly long method (I like short methods)

Reply:
(copying from slack)

Wonder if the quickest win would be changing it from a hash to a case statement, then it's not gonna generate all the paths. Will investigate

Reply:
I'd be happy with that.